lib/server.js

```diff
@@ -0,0 +1,198 @@
+            debug('paused');
+          }
+          // debug('server needs drain: %s',!tf);
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+
+    function resumed (message) {
+      var buf = new Buffer(new Uint8Array(message));
+      debug(buf);
+      dbgWrite('IN', buf);
+      switch (message[0]) {
+        case msgcode.pause:
+          strbuf.pause();
+          socket.removeListener('message', resumed);
+          socket.on('message', paused);
+          debug('paused');
+          break;
+
+        case msgcode.stdin:
+          if (!execStream.write(buf.slice(1), {binary: true})) {
+            strbuf.write(new Buffer([msgcode.pause]), {binary: true});
+            debug('paused');
+          }
+          // debug('server needs drain: ',!tf);
+          break;
+
+        default:
+          debug('unknown msg code %s', message[0]);
+          throw new Error('unknown msg code');
+      }
+    }
+    socket.on('message', resumed);
+
+    socket.on('disconnect', () => {
+      //should be how to ctrl+c ctrl+d, might be better way to kill
+      execStream.end('\x03\x04\r\nexit\r\n');
+    });
+
+    //start recieving client output again
+    execStream.on('drain', () => {
+      strbuf.write(new Buffer([msgcode.resume]), {binary: true});
+      debug('resumed');
+    });
+
+    //When the process dies, the stream ends
+    execStream.on('end', () => {
+      exec.inspect().then((data) => {
+        var exitBuf = new Buffer([data.ExitCode]);
+        strbuf.write(muxBuffer(msgcode.stopped, exitBuf));
+      }).done();
+    });
+
+    //send ready message to let client know it is ready
+    strbuf.write(new Buffer([msgcode.resume]));
+  }).done();
+}
+
+export default class DockerExecWebsocketServer {
+  /* Creates Docker Exec instance on given container, running the first message given
+   * as a command.
+   * Options:
+   * port, required
+   * containerId, name or id of docker container, required
+   * path, path where the websocket is hosted
+   * log, whether or not we want to log the i/o
+   */
+  constructor (options) {
+    options = _.defaults({}, options, {path: '/'+slugid.v4(), log: false});
+
+    //getting container
+    assert(options.containerId, 'required container option missing');
+    var container = docker.getContainer(options.containerId);
+    assert(container, 'could not get container from Docker');
+
+    this.port = options.port;
+    this.path = options.path;
+
+    //making websocket server
+    var wsopts = {
+      port: this.port,
+      path: this.path,
+    };
+    assert(this.port, 'required port option missing');
+    this.server = new ws.Server(wsopts);
+    debug('%s%s created', wsopts.port, wsopts.path);
+
+    //creating optional debug stream
+    var dbgStream;
+    if (options.log) {
+      dbgStream = fs.createWriteStream(path.join(__dirname, '/dump.log'));
```

Comment:
Bad idea... let people specify the log file... or even better let them specify a dbgStream.
This is certainly not good in a module...